### ISSUE TITLE
Countdown to the new Images.jl

### ISSUE THREAD

**Issue by timholy** [JuliaImages/Images.jl#542]
There have been justifiable questions about "what exactly is happening?" so let me post my TODO list here (including completed items, with an explanation of purpose):
- [X] ImagesCore.jl: handles low-level tasks like color separations, raw <--> ufixed conversions, dimension permutations. The new feature is the use of julia's swanky new array technology to provide views for all of these operations, so that we don't sometimes-copy-and-sometimes-return-a-view. Note this also effectively (partly in conjunction with the next package) solves the "Images is really confusing about image orientation") problem, becuase of the dimension permutation views.
- [X] ImagesAxes.jl: images with "named" dimensions, most especially temporal dimensions (like movies). Based on AxisArrays.jl, which also allows one to specify a `Range` (even one with physical units) for each axis, which thus incorporates the `pixelspacing`. All this information is exposed to the type system (rather than through fields of the properties dict). What that means is that indexing operations become type-predictable: for example, if you slice at a particular spot in time, the new array doesn't have a time axis anymore. In the old Images.jl we had to check for this and set the `timedim` property to 0 at runtime, and because that's slow we were constantly throwing away the dictionary for performance and therefore losing track of orientation. Now everything should Just Work.
- [X] ImagesMeta.jl: for those cases where one still wants to attach metadata to images.
- [x] ImagesFiltering.jl: better `imfilter` implementations. Everything will be type-stable, efficient algorithms will be chosen automatically based on the kernel size, and a platform will be in place for more efficient operations with separable kernels. Also options for performing the computation on the GPU, etc. This package will also likely be the first place to test another experimental direction: the ability (at the user's choice) to dispatch to algorithms in other libraries like OpenCV. (Currently this package is well underway but still needs work, and has been the occasion for me making changes to julia and a half dozen other packages. It won't be runnable by anyone else until julia-0.5-rc3 at the earliest.)
- [x] rename the above packages to ImageCore etc.
- [x] Get necessary dependencies registered and tagged. In addition to the above-mentioned packages, these include:
  - [x] AxisArrays
  - [x] SimpleTraits
  - [x] CatIndices
  - [x] FFTViews
  - [x] ComputationalResources
  - [x] Ranges (https://github.com/JuliaArrays/Ranges.jl) for Ranges that are safe with [Unitful](https://github.com/ajkeller34/Unitful.jl) quantities (cf https://github.com/JuliaLang/julia/pull/18777)
- [x] Replace the old `MapInfo` machinery with something based on functions. This will probably go in ImageCore.
- [x] Make sure that tests are good, and that as many deprecations are in place as possible.
- [x] Make sure all documentation is in place, and good:
  - [x] decide about "local" (detailed docs in each pkg) vs "centralized" (everything at juliaimages.io)
  - [ ] for each package, make `?pkg` give an API summary
  - [ ] edit the pkg descriptions (https://github.com/timholy/Images.jl/issues/542#issuecomment-254480264)
  - [ ] add usage examples (community contributions necessary here)
- [x] On a branch (let's call it "images-next"), rebase Images.jl on these packages. Will make heavy use of `@rexport`. Get the old tests passing:
  - [x] core.jl (src file will be deleted)
  - [x] map.jl
    - [x] src/map.jl needs deprecation strategy
  - [x] overlays.jl
  - [x] algorithms.jl
  - [x] exposure.jl
  - [x] edge.jl
  - [x] writemime.jl (needed for IJulia)
  - [x] corner.jl
  - [x] distances.jl
- [ ] Update IO packages:
  - [x] ImageMagick
  - [ ] QuartzImageIO
  - [x] Netpbm
  - [x] NRRD (will return AxisArray objects for axes with physical units or "meaning" to the axes)
- [x] Post an announcement here to say "check out branch images-next" if you want to play with the new Images" and collect comments from enterprising testers for about a week or so.
- [ ] Merge to master
- [ ] Collect comments for another week or so
- [ ] Tag a new release and post an announcement on julia-users

Things that could be done before or after:
- [ ] Possibly transfer all the intensity-transformation work to a separate package? Basically my thinking is that largish self-contained units might be a separate module. This would include all of @mronian's great work ~~as well as the `src/map.jl` functionality. (For now I don't really plan on touching that, so that file could probably be copied verbatim.)~~
- [ ] Resizing, rotation, restriction, etc. This seems like a self-contained chunk that should probably be a standalone package? I am not working on this right now at all, I would welcome contributions (@Evizero you have expressed interested in this area?)
- [x] Possibly split Overlay into a separate package

Longer-term future: implement lots of missing functions. Other **BSD-licensed** projects should be used as important inspiration.

**Comment by Evizero** [comment]
Absolutely! Right now I do the operation chaining for my image augmentation package **unnecessary** inefficient by creating temporary copies after each operation (input -> cropped image -> rotated image -> distorted image .... etc).

After watching your talk I am really sold on the idea of lazy views, which should - if I am not mistaken - allow for just nesting views instead; which would avoid my predicament.

That said, I am not an image processing or computer vision person, so bare with my newbie mistakes I am bound to make, but I am catching on.

To be a little less vague about my true intentions, I essentially want to create a bridge between julias nice Image functionality and machine learning needs, which diverge in some strange places. For example the "z" or "time" dimension does for us store observations with no intrinsic order/relation. Even stranger: in our usecases the color dimension of an image is actually farther apart in memory than the spatial!. so a 3x20x20x1000 image would often be represented in 20x20x3x1000 (this is just a consequence of how some of our algorithms work)

Anyway, thank you for taking the time to explain this!

**Comment by timholy** [comment]
With regards to your 20x20x3x1000 array, using ImagesCore you can already do this:
```jl
julia> using ImagesCore, Colors

julia> a = rand(UInt8, 20, 20, 3, 1000);

julia> summary(a)
"20×20×3×1000 Array{UInt8,4}"

julia> p = permuteddimsview(a, (3,1,2,4));

julia> c = colorview(RGB, ufixedview(p));

julia> summary(c)
"20×20×1000 ImagesCore.ColorView{ColorTypes.RGB{FixedPointNumbers.UFixed{UInt8,8}},3,MappedArrays.MappedArray{FixedPointNumbers.UFixed{UInt8,8},4,Base.PermutedDimsArrays.PermutedDimsArray{UInt8,4,(3,1,2,4),(2,3,1,4),Array{UInt8,4}},ImagesCore.##15#17{FixedPointNumbers.UFixed{UInt8,8}},ImagesCore.##16#18}}"

julia> col = c[5, 3, 80]
RGB{U8}(0.953,0.914,0.208)

julia> red(col).i
0xf3

julia> a[5,3,1,80]
0xf3

julia> green(col).i
0xe9

julia> a[5,3,2,80]
0xe9

julia> blue(col).i
0x35

julia> a[5,3,3,80]
0x35

julia> a[5,3,2,80] = 0
0

julia> c[5,3,80]
RGB{U8}(0.953,0.0,0.208)
```
That's the thing about these views...it takes a while to realize how something that seems so "boring" can be so insanely useful. In this case we put a "permuteddimsview" inside of a "ufixedview" inside of a "colorview" and should get darn good performance despite all those shennanigans.

Now the only thing left is to wrap it in an `AxisArray` (using ImagesAxes.jl) that indicates that the final dimension is time (if that's what you are using it for).

**Comment by mronian** [comment]
Maybe we could include ```corners```(recently overhauled) and ```edges```(probably needs changes) in ImageFeatures.jl since they come under feature detection.

:+1: for 100% test coverage ! Love to see the bright green coverage badge :smile: 

Regarding the documentation, I was thinking of having one ```juliaimages.github.io``` which will have the docs for everything. I am still a bit confused how to use Documenter.jl to tap into each repository and get the docstring. I think one of the other GSoCers has done it and I'll take help from him. Does this seem like a good idea or should there be a separate documentation for each one?

**Comment by mronian** [comment]
Should TestImages.jl be moved to JuliaImages? We could also have some other package (ImageSets.jl ?) which will have the famous datasets in a easily importable format.

**Comment by mronian** [comment]
Maybe we could have stuff like tutorials etc on the main website and if someone wants to dig into the function reference then they would be pointed to the individual websites.

**Comment by rsrock** [comment]
Before I read any of the above, let me say thanks for filling us in!  I was wondering "what exactly is happening???"

**Comment by rsrock** [comment]
All of the above sounds great, I can't wait to kick the tires.  

Here's one thing for my personal wishlist, a pretty printer for image types.  When I see something like:

```
"20×20×1000 ImagesCore.ColorView{ColorTypes.RGB{FixedPointNumbers.UFixed{UInt8,8}},3,MappedArrays.MappedArray{FixedPointNumbers.UFixed{UInt8,8},4,Base.PermutedDimsArrays.PermutedDimsArray{UInt8,4,(3,1,2,4),(2,3,1,4),Array{UInt8,4}},ImagesCore.##15#17{FixedPointNumbers.UFixed{UInt8,8}},ImagesCore.##16#18}}"
```

I have to sit there and count curly braces to figure out where I am! A few spaces in there would help as well.

**Comment by timholy** [comment]
@mronian,

> Maybe we could include corners(recently overhauled) and edges(probably needs changes) in ImageFeatures.jl since they come under feature detection.

I was wondering the same thing. I am in the middle of putting the gradient kernels into ImagesFiltering, but the edge-detection stuff doesn't quite fit. One option is to leave them in Images (src/algorithms won't entirely go away), but I tend to agree that ImageFeatures might be a better choice. CC @kmsquire to see if he has a preference.

> :+1: for 100% test coverage ! Love to see the bright green coverage badge :smile: 

Yes, that's important. The new repos are generally closer to that goal. ImagesAxes is much better than it looks, because some bug (probably in julia) incorrectly misses test coverage of traits-functions. See https://github.com/mauro3/SimpleTraits.jl/pull/6.

> Regarding the documentation, I was thinking of having one juliaimages.github.io which will have the docs for everything.

I've had the same thought. I think it will be good to have the docs for Images.jl take much more of a tutorial style, and they can link out to the docs for the individual packages for details. But it would be good to also have one giant function reference. I *think* that will be fairly straightforward with Documenter, but I haven't tested yet.

> Should TestImages.jl be moved to JuliaImages? We could also have some other package (ImageSets.jl ?) which will have the famous datasets in a easily importable format.

Yes, good idea.

@rsrock,

> Here's one thing for my personal wishlist, a pretty printer for image types. 

Yes, I agree. Ideally that would be something fixed in base julia, but perhaps we could pioneer something here. Jeff has been violently opposed to abbreviating the output of `typeof(x)`, but seems fine with customizing the printing of `x` itself. So we could conceivably modify `Base.summary`. We'd have to do it for all array types, or perhaps all array types of `Colorant`s, which is a little scary, because it would affect other packages too that were loaded at the same time as Images.

**Comment by kmsquire** [comment]
I think putting edge detection in ImageFeatures would be fine.

**Comment by mronian** [comment]
When we move an algorithm say A, how do we do so without affecting either of the packages? If I just directly copy, it gives a redefinition of A error. What is the correct way to do this so that A is ported and none of the packages depending on the older A are affected?

**Comment by Evizero** [comment]
>  I think it will be good to have the docs for Images.jl take much more of a tutorial style

@timholy What do you have in mind here? I wrote a beginner's tutorial to julia's approach to handling image data for a project of mine [over here](http://augmentorjl.readthedocs.io/en/latest/tutorials/images.html), and I'd happily contribute it if you want it (it is somewhat WIP still). But since it is mainly aimed at ML students (possibly new to Julia), and non-image-experts, I am guessing it might be a little too beginner oriented to be useful to you? (also, feel free to criticize me when I made a mistake or over/understated something)

**Comment by timholy** [comment]
> When we move an algorithm say A, how do we do so without affecting either of the packages?

I think we have to do merges to master/tagging in synchrony. In other words, start a new branch in both packages that make the same change.

> I wrote a beginner's tutorial 

I haven't read it in detail, but this looks like exactly the kind of thing I was envisioning. Most will probably display images in the docs, too.

I haven't looked in detail, but http://scikit-image.org/docs/dev/user_guide.html seems like a promising model to follow.

**Comment by Evizero** [comment]
where would you want such tutorials to be located at?

**Comment by timholy** [comment]
I think Images.jl should be the home of all the "major" documentation. (Images.jl will move to JuliaImages at the time of the transition.)

**Comment by mronian** [comment]
We could have the docs in a separate repository -> https://github.com/JuliaImages/juliaimages.github.io so that juliaimages.github.io becomes the website.

**Comment by Evizero** [comment]
One thing to keep in mind is that if the docs are to contain lots of images, is that they have to be stored somewhere, which can be a hassle depending on where the docs are (since one wants to keep the package size small and the git history clean)

**Comment by mronian** [comment]
We faced this problem in TestImages.jl for the documentation. Turns out you can use the ```gh-pages``` branch to store all website related stuff. :)

**Comment by timholy** [comment]
All good suggestions. :+1: to putting it in the org.

**Comment by ViralBShah** [comment]
Is `JuliaImages` going to be the new home for all this?

**Comment by ViralBShah** [comment]
We should certainly do a blog post on julialang.org once this is done.

**Comment by tlnagy** [comment]
Will the new `Images` be able to be installed and used with a simple `using Images`? Because right now it looks like it is broken into a huge number of smaller packages, which could be a nightmare to interface with.

**Comment by timholy** [comment]
Yes, it will. First, they will all be REQUIREd packages, so they'll be installed automatically. Second, their functions will all be [`@reexport`ed](https://github.com/simonster/Reexport.jl), so unless you want just a subset of the functionality of `Images` (and therefore load specific sub-packages) you won't even have to care how the functionality is partitioned at the source-code level. Images will probably still have some "real" code in it, but in many ways it is evolving into a meta-package that just bundles a lot of related functionality together.

This split-up is partly designed to make them more accessible to people who want a subset of the functionality of Images...for example https://github.com/JuliaLang/julia/issues/18384. There is no way to satisfy the desire to "reuse code across different types of applications" unless we split things into relatively focused units of functionality. Reexporting is the key that allows you to have this without sacrificing on other equally-worthy goals, like "provide a comprehensive Images package."

**Comment by tlnagy** [comment]
Thanks for the clarification. That sounds promising, I haven't heard of
the @reexport feature yet. Is there a framework for running integration
tests in the meta package?

**Comment by timholy** [comment]
You can run any tests you want...but it sounds like you have something specific in mind with "framework", what do you have in mind?

**Comment by tlnagy** [comment]
So currently most Julia packages have a test/ folder with unit tests
that gets run every time a new commit is pushed to master or a PR is
updated. I guess each subpackage of Images will load all of Images every
time and test all functionality to make sure that there are no
regressions in some other Images subpackage. Now that I said it, it
sounds reasonable.


On Wed, Sep 7, 2016, at 15:06, Tim Holy wrote:
> You can run any tests you want...but it sounds like you have something
> specific in mind with "framework", what do you have in mind?
> — You are receiving this because you commented. Reply to this email
> directly, view it on GitHub[1], or mute the thread[2].
>

Links:

  1. https://github.com/timholy/Images.jl/issues/542#issuecomment-245434736

**Comment by mronian** [comment]
We could define the `runtests.jl` in Images.jl in such a way that it provides a few options like running specific tests or the tests of each package.

**Comment by timholy** [comment]
I see your point. This is a valid issue to be concerned about.

In general I would go the other way: I would have the focused packages test their own functionality, and have the meta-package test everything. You're right, though, that this means it's possible to inadvertently break packages, because if you're updating one of the "focused" packages it won't automatically trigger the Images tests.

One model is to have extra tests that run only on Travis. For example, JLD tests JLDArchives (a repository of "outdated" `*.jld` files) to ensure backward-compatibility. But this doesn't run as part of `Pkg.test("JLD")` or `include("runtests.jl")` for just JLD on your local machine--it's something that is specially triggered in the `.travis.yml` file.

So using this model, each package could have its own focused tests, but then the `.travis.yml` script also forces the Images.jl tests to run too, to make sure the PR doesn't break other packages. The one downside is that CI will be slower, but perhaps that's the price we have to pay.

**Comment by tlnagy** [comment]
Agreed. I think it would still be nice to add an option to `runtests` like @mronian said because otherwise it would be difficult to test your changes locally prior to pushing the PR. It shouldn't be that much slower than the running the full test suite of the current `Images.jl` because it's the same number of tests, but split into smaller containers.

**Comment by mronian** [comment]
Will `Images.jl` have any functionality itself apart from being a meta package (Since ImageCore etc are already separated) ?

**Comment by tlnagy** [comment]
It's already on @timholy's todo list: 

> On a branch (let's call it "XYZ"), rebase Images.jl on these packages. Will make heavy use of @rexport.

**Comment by mronian** [comment]
If it itself doesn't have any functionality, will the tests be similar to those in individual packages? I mean, do we need separate tests again for `Images.jl` or will the individual package tests suffice?

**Comment by tlnagy** [comment]
> I mean, do we need separate tests again for Images.jl or will the individual package tests suffice? 

It would be nice to have a quick way to run all tests of `Images*` (`ImagesFiltering` etc) in one base. I envision something like this: make small tweak in `ImagesFiltering` and run `Pkg.test("Images")` and make sure you didn't cause any regressions in the other tightly linked `Images*` packages.

**Comment by mronian** [comment]
I hope we find a workaround :sweat_smile: as CI will quickly become very slow as these packages are very heavy (ImageFeatures.jl itself takes around 20-30 min now and with additional functionality like AGAST the tests will only bloat up ) :disappointed:

**Comment by tlnagy** [comment]
> Here's one thing for my personal wishlist, a pretty printer for image types. When I see something like:

> ```
20×20×1000 ImagesCore.ColorView{ColorTypes.RGB{FixedPointNumbers.UFixed{UInt8,8}},3,MappedArrays.MappedArray{FixedPointNumbers.UFixed{UInt8,8},4,Base.PermutedDimsArrays.PermutedDimsArray{UInt8,4,(3,1,2,4),(2,3,1,4),Array{UInt8,4}},ImagesCore.##15#17{FixedPointNumbers.UFixed{UInt8,8}},ImagesCore.##16#18}}"
```

> I have to sit there and count curly braces to figure out where I am! A few spaces in there would help as well.

I would go further than @rsrock and say that this is basically unreadable as is. I can only imagine what error messages look like with that type printed everywhere.

**Comment by timholy** [comment]
I also find these extremely difficult to read, and would love to change this. The key question is how to fix it.

Currently we rely on the generic [`Base.summary`](https://github.com/JuliaLang/julia/blob/155147e46c87300bf236c8dc12b2761c59441ac1/base/show.jl#L1516-L1518) for printing that kind of type information. The impenetrably-nested braces comes from that `typeof(a)`. Now, we're [advised against changing how types are printed](https://github.com/JeffBezanson/FixedPointNumbers.jl/pull/9#r16670933), which is bad news for addressing this:

> I can only imagine what error messages look like with that type printed everywhere.

But it's legitimate to provide a specialized `show` or `summary` function, since that's used for value-printing. So we can at least fix some aspect of this.

I guess in my mind the key question is, how to specialize it? I see three options:

1. override the Base definition for every `AbstractArray`
2. specialize for arrays that have a `Colorant` eltype
3. come up with a list of containers (at least ColorView, ChannelView, MappedArrays, ImageMeta) that deserve specializations of `summary` (if current plans hold, there will be no "one container to rule them all" when it comes to images).

Pros and cons:
1. Pro: could perhaps really fix the issue. Current thought would be to compute the nesting depth and then do something immensely clever (just have to figure out what that is :wink:) depending on the answer. Con: this will cause a warning to be printed every time you say `using Images`. More subversively, it would affect the behavior of any other package that the user had co-loaded with Images---it could mess up printing for their (array) types. That's considered bad form.
2. Pro: most images will presumably be loaded as Colorant types. In particular, IJulia will only show arrays that have Colorant eltypes as an image; the rest will fall back to text output. Con: leaves out `ChannelView` and any grayscale images that don't explicitly use `Gray`.
3. Pro: probably the cleanest in terms of encapsulating for types we care about. Con: will miss important types like `SubArray` and `PermutedDimsArray`, which are also very useful containers for images.

To be honest, my favorite solution is 1. I think we could take something for a test drive here, knowing that we're committing a faux pas, with the excuse that if we all like it we'll submit it as a PR for Base.

This is the moment where somebody reading this chimes in with the brilliant suggestion that gets the best of all possible worlds.

**Comment by timholy** [comment]
While on the topic of how things are printed, perhaps it's worth pointing out https://github.com/JeffBezanson/FixedPointNumbers.jl/pull/51#issuecomment-247507435, which is a proposal to change how the fixed-point types are printed. A major motivation for this proposal is to make text-display of pixels values (or whole images) more compact.

**Comment by timholy** [comment]
I've periodically been updating the list above. It feels like it's getting close now.

There will be some delay, unfortunately. I did a fair amount of work updating ImageMagick/writemime, but then my laptop died (motherboard problem), and my old laptop wouldn't read the new ones' hard drive. So some items will likely wait until I get my laptop back. So far I haven't found myself short of other tasks in that list to keep me occupied :persevere: .

**Comment by timholy** [comment]
Locally I have a candidate solution to the printing-of-types problem mentioned above as being high on @rsrock's wish list, https://github.com/timholy/Images.jl/issues/542#issuecomment-239647238. I figured out a way to do it that will only affect types defined in the images-related packages (I think).

The idea is based on the observation that sometimes the type printing is more confusing than the sequence of calls you'd use to create the object. So I'm trying out a "show it like you build it" approach (I even have a nascent package with that name). Here is a demo:
```jl
julia> a = rand(UInt8, 5,5,3)
5×5×3 Array{UInt8,3}:
[:, :, 1] =
 0x63  0x44  0xe1  0x1e  0xba
 0x08  0x07  0x3f  0x4c  0x1f
 0x10  0xaf  0xf7  0xe7  0x87
 0x3c  0x37  0x49  0x6e  0xd4
 0xd9  0x05  0xf2  0x5b  0x0f

[:, :, 2] =
 0x8a  0xeb  0x34  0x6b  0x59
 0x0f  0xd6  0x71  0x8d  0x74
 0xe6  0xfa  0xb1  0xba  0xeb
 0x9b  0xb6  0xa4  0xe2  0x24
 0xe2  0x4f  0x62  0xc5  0x70

[:, :, 3] =
 0xfd  0x31  0x24  0x49  0xf3
 0x9a  0x2d  0x80  0x2c  0x10
 0xeb  0x0e  0x03  0x68  0x68
 0xbd  0x0e  0x39  0xf0  0x0a
 0xc1  0x0c  0x66  0x49  0x78

julia> p = permuteddimsview(a, (3,1,2))
3×5×5 permuteddimsview(::Array{UInt8,3}, (3,1,2)) with element type UInt8:
[:, :, 1] =
 0x63  0x08  0x10  0x3c  0xd9
 0x8a  0x0f  0xe6  0x9b  0xe2
 0xfd  0x9a  0xeb  0xbd  0xc1

[:, :, 2] =
 0x44  0x07  0xaf  0x37  0x05
 0xeb  0xd6  0xfa  0xb6  0x4f
 0x31  0x2d  0x0e  0x0e  0x0c

[:, :, 3] =
 0xe1  0x3f  0xf7  0x49  0xf2
 0x34  0x71  0xb1  0xa4  0x62
 0x24  0x80  0x03  0x39  0x66

[:, :, 4] =
 0x1e  0x4c  0xe7  0x6e  0x5b
 0x6b  0x8d  0xba  0xe2  0xc5
 0x49  0x2c  0x68  0xf0  0x49

[:, :, 5] =
 0xba  0x1f  0x87  0xd4  0x0f
 0x59  0x74  0xeb  0x24  0x70
 0xf3  0x10  0x68  0x0a  0x78

julia> u = ufixedview(p)
3×5×5 ufixedview(N0f8, permuteddimsview(::Array{UInt8,3}, (3,1,2))) with element type FixedPointNumbers.UFixed{UInt8,8}:
[:, :, 1] =
 0.388N0f8  0.031N0f8  0.063N0f8  0.235N0f8  0.851N0f8
 0.541N0f8  0.059N0f8  0.902N0f8  0.608N0f8  0.886N0f8
 0.992N0f8  0.604N0f8  0.922N0f8  0.741N0f8  0.757N0f8

[:, :, 2] =
 0.267N0f8  0.027N0f8  0.686N0f8  0.216N0f8  0.02N0f8 
 0.922N0f8  0.839N0f8  0.98N0f8   0.714N0f8  0.31N0f8 
 0.192N0f8  0.176N0f8  0.055N0f8  0.055N0f8  0.047N0f8

[:, :, 3] =
 0.882N0f8  0.247N0f8  0.969N0f8  0.286N0f8  0.949N0f8
 0.204N0f8  0.443N0f8  0.694N0f8  0.643N0f8  0.384N0f8
 0.141N0f8  0.502N0f8  0.012N0f8  0.224N0f8  0.4N0f8  

[:, :, 4] =
 0.118N0f8  0.298N0f8  0.906N0f8  0.431N0f8  0.357N0f8
 0.42N0f8   0.553N0f8  0.729N0f8  0.886N0f8  0.773N0f8
 0.286N0f8  0.173N0f8  0.408N0f8  0.941N0f8  0.286N0f8

[:, :, 5] =
 0.729N0f8  0.122N0f8  0.529N0f8  0.831N0f8  0.059N0f8
 0.349N0f8  0.455N0f8  0.922N0f8  0.141N0f8  0.439N0f8
 0.953N0f8  0.063N0f8  0.408N0f8  0.039N0f8  0.471N0f8

julia> c = colorview(RGB, u)
5×5 ColorView{RGB}(ufixedview(N0f8, permuteddimsview(::Array{UInt8,3}, (3,1,2)))) with element type ColorTypes.RGB{FixedPointNumbers.UFixed{UInt8,8}}:
 RGB{U8}(0.388,0.541,0.992)  RGB{U8}(0.267,0.922,0.192)  RGB{U8}(0.882,0.204,0.141)  RGB{U8}(0.118,0.42,0.286)   RGB{U8}(0.729,0.349,0.953)
 RGB{U8}(0.031,0.059,0.604)  RGB{U8}(0.027,0.839,0.176)  RGB{U8}(0.247,0.443,0.502)  RGB{U8}(0.298,0.553,0.173)  RGB{U8}(0.122,0.455,0.063)
 RGB{U8}(0.063,0.902,0.922)  RGB{U8}(0.686,0.98,0.055)   RGB{U8}(0.969,0.694,0.012)  RGB{U8}(0.906,0.729,0.408)  RGB{U8}(0.529,0.922,0.408)
 RGB{U8}(0.235,0.608,0.741)  RGB{U8}(0.216,0.714,0.055)  RGB{U8}(0.286,0.643,0.224)  RGB{U8}(0.431,0.886,0.941)  RGB{U8}(0.831,0.141,0.039)
 RGB{U8}(0.851,0.886,0.757)  RGB{U8}(0.02,0.31,0.047)    RGB{U8}(0.949,0.384,0.4)    RGB{U8}(0.357,0.773,0.286)  RGB{U8}(0.059,0.439,0.471)
```

So this still shows as nested, but I think
```jl
ColorView{RGB}(ufixedview(N0f8, permuteddimsview(::Array{UInt8,3}, (3,1,2)))) with element type ColorTypes.RGB{FixedPointNumbers.UFixed{UInt8,8}}
```
is easier to read than
```jl
julia> typeof(c)
ImageCore.ColorView{ColorTypes.RGB{FixedPointNumbers.UFixed{UInt8,8}},2,MappedArrays.MappedArray{FixedPointNumbers.UFixed{UInt8,8},3,Base.PermutedDimsArrays.PermutedDimsArray{UInt8,3,(3,1,2),(2,3,1),Array{UInt8,3}},ImageCore.##29#30{FixedPointNumbers.UFixed{UInt8,8}},Base.#reinterpret}}
```

Thoughts? Is this a good direction? If so, is it sufficient, or are there adjustments needed?

**Comment by Evizero** [comment]
I like it. certainly easier than the type. I am not quite sure I fully follow the pattern yet. For example why write `ColorView{RGB}`, but `unfixedview(N0f8,` when both originated from a similar looking method call

**Comment by timholy** [comment]
To explain: it's because `ColorView{C}(A)` *always* creates a `ColorView`, whereas `colorview(C, A)` *might* create a `ColorView`. It just returns a plain array when `A` is amenable to `reinterpret`. This strategy is an attempt to reduce the amount of nesting in the types themselves.

I agree that's a little confusing, though, so I'm happy to entertain alternatives.

**Comment by Evizero** [comment]
Isn't that kind of dispatch true for `ufixedview` as well https://github.com/JuliaImages/ImageCore.jl/blob/master/src/ImageCore.jl#L90? I guess I am just unclear when a function-style printing is chosen (such as for `ufixedview`) vs when a type-style printing is chosen (such as for `colorview`)

Regardless of my confusions (which are numerous no matter what anyways), I do like the change. 👍

**Comment by timholy** [comment]
I would call this "productive confusion," in the sense that this is 100% valid. So no apologies needed; to the contrary, thanks for raising these concerns.

In the case of `ufixedview`, I'm not certain I see a way to resolve the issue without introducing problems elsewhere:

- Let's say you always want to show it as a type. But there is no type directly corresponding to `ufixedview`: both `ufixedview` and `rawview` create a [MappedArray](https://github.com/JuliaArrays/MappedArrays.jl) (when `reinterpret` wouldn't work). We could print it as `MappedArray((x->reinterpret(Nf08, x), reinterpret), <inner array type>)` but I worry that's considerably harder to read than `ufixedview(Nf08, <inner array type>)`. (But this is nevertheless a viable option.) Alternatively, we could introduce `UFixedView` and `RawView` types, but that's an awful lot of duplication of functionality for no actual benefit when MappedArrays already exists.

- Let's say you always want to show it as a function call. Then the problem is that you *could* create a `ColorView` like this: `ColorView{RGB}(rand(3,5,5))`. Then in this scenario that would print as `colorview(RGB, ::Array{Float32,3})`, but the problem is that `colorview(RGB, rand(3,5,5))` would return an `Array` and not a `ColorView`, which is the actual type of the object that's being printed.

**Comment by shorty66** [comment]
As there are alternative IO-solutions available now, is it still mandatory to install **ImageMagick**? 
Im asking as IM is blocked at my company due to [serious security flaws](https://imagetragick.com/).
It would be nice to have a query at installation to select one of the IO-packages.

**Comment by timholy** [comment]
There is a query on OSX because of [QuartzImageIO](https://github.com/JuliaIO/QuartzImageIO.jl), but there isn't a comprehensive alternative on any other platform. [Netpbm](https://github.com/JuliaIO/Netpbm.jl) is probably closest.

This is a great area for motivated folks to contribute! If you have another library in mind, you can see from ImageMagick.jl that it's not an outrageously large undertaking to write a wrapper. (Not a trivial amount of work, either.) Or more ambitiously one could write a parser in pure julia (like the Netpbm and NRRD parsers). But that would be one format at a time, and a much larger undertaking.

**Comment by Evizero** [comment]
http://www.graphicsmagick.org/index.html

**Comment by timholy** [comment]
See also previous discussion in https://github.com/timholy/Images.jl/issues/24

**Comment by shorty66** [comment]
Actually, [GraphicsMagick](https://sourceforge.net/p/graphicsmagick/mailman/message/35072963/) is known to have inherited some of the security flaws from ImageMagick. For that reason, it is also blocked at our company.
I will see if i can write a little wrapper for some basic IO library.

**Comment by timholy** [comment]
## Near the end of a long journey

I'm pleased to announce that the "new Images" is ready for testing. It's not that absolutely everything is done (see below), but I think it would be appropriate for people to start playing with this. For a list of some of the main changes, please see https://github.com/timholy/Images.jl/blob/images-next/NEWS.md.

### Documentation

I've put a lot of work into documentation so far, although that effort isn't completely done. The new documentation will be hosted mostly at http://juliaimages.github.io/latest/, which is intended to be "organizational-level" documentation for all of JuliaImages. Before working on that documentation, I'd also written documentation for ImageCore, ImageFiltering, ImageAxes, and ImageMetadata---I'm wondering if this should just be moved to JuliaImages?

I am hoping this will become very polished documentation, competitive with other image processing suites like skimage, OpenCV, etc., but it is not there yet. What *is* there is supposed to be good, so do read it with a critical eye and please fix/report problems. In my mind the "tutorial" sections are mostly complete, but it's missing important components (like API documentation) and most especially a bunch of examples (e.g., see http://scikit-image.org/docs/dev/auto_examples/index.html). I **really** hope that users will contribute to the documentation especially by providing examples. 

### Trying it out

```jl
Pkg.update()
```
Then download [this gist](https://gist.github.com/timholy/604ffd44bdb76140ed17c2ba160bb6c1) (or just copy/paste the file), and say
```jl
include("imagebranch.jl")
imagebranch("fixed-renaming")
```
**If you have dirty packages or are on branches, this will almost certainly fail.** I have verified that it does work with a fresh pkg folder, e.g.,
```jl
tim@cannon:~/src/julia-0.5$ JULIA_PKGDIR=/tmp/tim/pkgs2 ./julia -q
julia> Pkg.init()
INFO: Initializing package repository /tmp/tim/pkgs2/v0.5
INFO: Cloning METADATA from https://github.com/JuliaLang/METADATA.jl

julia> Pkg.add("PkgDev")
<some output>

julia> Pkg.add("Images")
<some output>

julia> Pkg.add("ImageMagick")  # plus any others you might need
<some output>

julia> include("/tmp/tim/imagebranch.jl")
INFO: Precompiling module PkgDev.
imagebranch (generic function with 1 method)

julia> imagebranch("fixed-renaming")
<lots of output>
```


If you find yourself unhappy and want to go back to the original, use
```jl
imagebranch("release")
```
and if you want to try the "intermediate" set (one that doesn't include https://github.com/JuliaMath/FixedPointNumbers.jl/pull/51), use
```jl
imagebranch("images-next")
```

### Stuff I'd like feedback on

- do you like (or hate) the new FixedPointNumber names?
- any bolded comments in the JuliaImages documentation
- are any deprecations missing? Some things (e.g., setting the "spatialorder" property) have to be errors, but does the error message give clear instructions about how to fix things?
- any problems (or things you especially like) about the API changes

Please report problems as separate issues (don't just add to this one), but mark it clearly as to which "overarching branch" you're using ("release", "images-next", or "fixed-renaming").

### Tips while running old scripts with new code

First, I urge you to **use a branch when making any changes to your own internal code**; who knows how long it will be before this merges to master, so you might want to be able to easily go back and forth.

Second, don't forget that you can launch julia as `julia --depwarn=no` and suppress a whole sea of deprecation warnings. That said, the point of the deprecation warnings is to teach you how to migrate your code, so you'll want to do at least enough of this to start discovering problem points, things you like or dislike, etc.

**Comment by Evizero** [comment]
Thanks! I was able to update using your gist, although I did have to visit all affected package folders and do `git fetch --all` (which may be an overkill, I know). Otherwise the command `imagebranch("fixed-renaming")` did not infer which corresponding remote branch to track and threw an error for each package I did not fetch

```
INFO: Checking out Colors fixed-renaming...
ERROR: GitError(Code:ERROR, Class:Merge, There is no tracking information for the current branch.)
[...]
```

**Comment by timholy** [comment]
Thanks! I updated the gist, now it should fetch automatically.

**Comment by timholy** [comment]
One thing I forgot to ask: do folks think that all the documentation should be moved to http://juliaimages.github.io/latest? (Meaning, should ImageCore, ImageFiltering, etc. continue to maintain their own documentation?) Either way, http://juliaimages.github.io/latest/conversions_views.html#Conversions-vs.-views-1 is largely redundant with https://juliaimages.github.io/ImageCore.jl/latest/views.html#Views-1. Any thoughts on this would be appreciated. Originally I intended to just link the page from ImageCore in to juliaimages.io, but there are [technical difficulties](https://github.com/JuliaDocs/Documenter.jl/issues/319) with that, and once I thought about just copying I came to the conclusion that some of the groundwork laid on earlier pages led to me wanting to explain things slightly differently. So if you think we should have only one of them, concrete changes for suggestions about which to keep or how to merge them would be appreciated.

**Comment by Evizero** [comment]
Personally I think the following would make sense:

- JuliaImages: Contains a high-level documentation of all the functionality without going into details (such as arithmetic overflow). I think this would also be a good location for this example tutorials that you envision. It should contain only that information that a user needs to work with his/her image data without the "burden" of technical details. I think the maturity of a library can often be inferred by how little a user has to learn in order to work with it, even if it means that the user remains oblivious to how cleverly some issues were addressed under the hood.

- The Packages: Detailed information and developer information for the provided functionality, such as `colorview` vs `ColorView`, the note on arithmetic overflow etc. All the information that makes the interested and potential contributors understand how the packages work.

**Comment by timholy** [comment]
The principle seems sound. But sometimes the dividing line between the two is unclear, perhaps especially to me (given that I spent so much time thinking about the technical details). To pick just one example: is http://juliaimages.github.io/latest/conversions_views.html#Sharing-memory:-an-introduction-to-views-1, pointing out the differences between a copy and and view, a "detail" or "necessary"?

My favored approach would be for others to cut/paste the docs and move them appropriately, since others have a better "outsider" perspective, but I'm happy to do it particularly if folks give me sufficiently detailed instructions.

**Comment by tbreloff** [comment]
My vote (not that I have one to cast) would be to have to docs all in one
place and dedicate a page to each package to explain how it fits into the
ecosystem and any peculiars one would need to know if they were to use it
in isolation. Then you could link directly to that page from the respective
repo.

**Comment by timholy** [comment]
You definitely have a vote to cast!

**Comment by tbreloff** [comment]
:)

Also... I could throw together a meta spec (for MetaPkg) for the transition
if you want, to avoid making users copy/paste that long gist to setup. I
want to help, but weary of pushing it on anyone considering the tension
over package management.

**Comment by timholy** [comment]
I haven't played with MetaPkg yet. Very cool, and seems like a systematic solution to precisely the problem my gist was designed to address. I'd be thrilled!

**Comment by tbreloff** [comment]
Exactly. My needs were for managing the many packages of JuliaPlots,
JuliaML, and the GLVisualize universe, many of them not registered.

**Comment by bjarthur** [comment]
would be nice to have more info in each package's README.  i code in a terminal and it's nice when packages contain at least a list of exports therein so that one can just say `?JLD`, for example, for a quick jog of the memory.

it's also very important to make the "description" (eg for ImagesCore it is currently "Julia types for representing images") informative.  because, this and the name of the package are all that's listed at http://pkg.julialang.org.  in such a constrained space, i think the word "Julia" is not useful.

**Comment by tlnagy** [comment]
I agree with @tbreloff here. One centralized place for documentation would be better. That way you don't have jump back and forth between different sets of docs. I would imagine that the less users are exposed to separate Image packages the better.

**Comment by tbreloff** [comment]
I took the liberty of adding a `JuliaImages` meta spec in MetaPkg. I tested it on my mac laptop, and everything seemed to work, but you should verify yourself (make sure all your work is pushed up because purge will do `Pkg.rm` and also delete cache/lib/trash files that are left around).

Instead of using that gist linked to above, you should be able to update with the commands:

```julia
using MetaPkg
MetaPkg.purge("JuliaImages")
meta_add("JuliaImages")
```

Then you should be all set up on the "release" path of the gist.

If you want to do the "fixed-renaming" branch:

```julia
meta_checkout("JuliaImages", "fixed-renaming")
```

and if you want to go back to the "release" afterwards:

```julia
meta_free("JuliaImages")
```

I didn't do the "images-next" path yet.. figured I'd get your input first.  I made [a gist](https://gist.github.com/tbreloff/7823919f1059d042872a4a1c3f990149) with the full expected output from the REPL.

**Comment by timholy** [comment]
This looks awesome.

I have a bunch of not-pushed branches in quite a few of those packages, and I'm hopping on a plane this morning, so it might take me a bit of time to prepare the groundwork for testing this. But I really like how this looks as a way of managing an interwoven set of packages!

**Comment by tknopp** [comment]
I am a little late on this show, but would like to ask where all the metadata handling has gone (`pixelspacing` and so on) This was a very prominent feature in the stable branch of Images and in our internal codebase almost anything is an image object for exactly that reason. We also attach quite some additional metadata in image objects.

**Comment by rsrock** [comment]
Tim, would you mind opening an issue over on QuartzImageIO with some guidance on what needs to be updated?  I haven't had a chance to try this out yet, but I hope to do so soon.  Thanks.

**Comment by timholy** [comment]
@tknopp, it's now called `ImageMeta`. See [ImageMetadata.jl](https://github.com/JuliaImages/ImageMetadata.jl). However, things like `pixelspacing` are no longer encoded in the dictionary; you should use [AxisArrays](https://github.com/JuliaArrays/AxisArrays.jl) (with a bit of extra functionality provided by [ImageAxes](https://github.com/JuliaImages/ImageAxes.jl)). All these packages are loaded when you say `using Images`, however. I'm especially interested in transitions like the one you're describing, so if you have more than a little difficulty with the transition please let me know: the whole point of this beta period is to find out the pain points, and I can help smooth the pain.

@rsrock, will do.

**Comment by tknopp** [comment]
Ok, thanks. Not sure when we get to this but we will try. I ping my colleague @hofmannmartin to be aware of the Images transition. Since we also set the pixelspacing using `im[:pixelspacing]` it seems that we will need some transitioning.

**Comment by timholy** [comment]
The safer thing has always been to say `pixelspacing(im)`, but perhaps this wasn't emphasized strongly enough.

**Comment by tlnagy** [comment]
@timholy I see you that you might move the `Overlay` stuff out. Would that be the recommended way of handling multichannel images or should something else be used instead?

**Comment by tknopp** [comment]
That certainly has been emphasized. There are these annoying people not carefully reading the documentation... Sorry my workflow is sometimes a little exploratory.

**Comment by timholy** [comment]
@tlnagy, that's in fact already done, see http://juliaimages.github.io/latest/conversions_views.html#Using-colorview-to-make-color-overlays-1 and redundantly described in https://juliaimages.github.io/ImageCore.jl/latest/views.html#View-types-defined-in-ImageCore-1. That allows you to easily fill in channels for red, green, and blue (and you can do magenta/green with `colorview(RGB, m, g, m)`). If that's not flexible enough, I suspect that better than the old approach would be to define a custom color type, e.g., the 2-channel color
```jl
immutable GFPtdTomato{T} <: Color{T, 2}
    gfp::T
    tdTomato::T
end
```
along with a function that can convert such "colors" to `RGB`.

**Comment by timholy** [comment]
@tknopp, now I also see you said "*set* the pixelspacing". Yes, it might be a little more work, sorry---if you find it too annoying then let's talk about a convenience function, e.g.
```jl
im = imageaxes(buf, pixelspacing=(1mm, 1mm))
```
or something.

As consolation for any pain, among the benefits of having it be more "structured" now is that it won't be lost when you index like `im[5:100, 40:80]`; formerly that lost both the pixel spacing and the orientation, which was just evil.

**Comment by tlnagy** [comment]
Thanks! The ImageCore documentation is quite thorough. It does mention `StackedView` on several occasions, but I can't seem to find any docs on that.

**Comment by timholy** [comment]
The balance of advice above was to centralize the docs at JuliaImages, so the final product may look more like what's there than the ImageCore docs.

`StackedView` is kinda internal; if you use `colorview(RGB, r, g, b)` then you don't need to touch it. Briefly, `StackedView` takes `m×n` arrays `a`, `b`, `c` and creates a `3×m×n` view of them.

**Comment by tknopp** [comment]
Is there some documentation how to combine `ImageMeta` objects with `AxisArrays`? I would like to start with an object that is a close to the current `Image` object and start from there.

**Comment by timholy** [comment]
Sure, they're completely orthogonal: just use an `AxisArray` for the array argument of `ImageMeta`. There's already some special-purpose code in place to look for this case, e.g., https://github.com/JuliaImages/ImageMetadata.jl/blob/b8e5c162b71babe457b56669e4cb8530964407f3/src/ImageMetadata.jl#L71-L73.

**Comment by tknopp** [comment]
So, a typical usage of Images.jl was

https://github.com/simonster/NIfTI.jl/blob/2f217aac10d0e48e77eeb437e82a11c203dc2255/examples/mriview.jl#L9

What is your advice to handle this in the future. I will try playing around with this myself, but maybe it would be good to have this particular case discussed in the documentation somewhere.

And whats about the other image loader, are the already ported to the new Image infrastructure?

**Comment by timholy** [comment]
In case you haven't seen it, https://juliaimages.github.io/ImageAxes.jl/latest/ might help.

Simple (ignoring pixel spacing): `ImageMeta(AxisArray(colorview(Gray, ni.raw), :x, :y, :z, :time))`

There isn't quite as simple an approach for handling the pixel spacing (this is *exactly* the kind of feedback I'm looking for), but it is more flexible:
```jl
ImageMeta(AxisArray(colorview(Gray, ni.raw), Axis{:x}(rngx), Axis{:y}(rngy), Axis{:z}(rngz), Axis{:time}(rngt)))
```
where `rng[xyzt]` is a range that describes the positions of the different coordinates, e.g., `0:2:284` if the pixels are 2mm and go from 0mm to 284mm. (If you want to attach real units, currently Julia's ranges are kinda problematic, but see [Unitful.jl](https://github.com/ajkeller34/Unitful.jl), [Ranges.jl](https://github.com/JuliaArrays/Ranges.jl) and https://github.com/JuliaLang/julia/pull/18777.)

Do you need an easier way to specify pixel spacing? The default `AxisArray` behavior is very flexible (it lets you specify the origin) but maybe somewhat verbose.

**Comment by timholy** [comment]
Note you don't really need the `ImageMeta` here, everything is encoded by the `AxisArray`...one of the main points of the redesign was to encode everything having to do with orientation in the type system so that `A[:, 5, :]` can preserve the names of the correct axes without hammering performance.

**Comment by tknopp** [comment]
Hm, this is of course flexible but also pretty complicated. 
The pixelspacing is one of the most important things since in our (tomographic) imaging method we have anisotropic voxels. Furthermore I use two custom fields "offset" and "rotation" within the image object that allow to define the precise coordinate system in space. With this I can take two `Image` objects measured with two different tomographs and overlay them doing the appropriate registration/regridding/image fusion. The rotation field is necessary since MRI scans can be arbitrarily angulated.

**Comment by tknopp** [comment]
So if I see it correctly I will have to but the `offset` into the ranges, but the `rotation` will remain an `ImageMeta` field.

**Comment by timholy** [comment]
Well, `offset` is exactly what you can encode in the `AxisArray` ranges. Might want a simpler approach, how about
```jl
A = AxisImage(data, (:x, :y, :z, :time), (dx, dy, dz, dt), [(x0, y0, z0, t0)])
```
?

As for rotation goes...that would either have to be in the metadata, or possibly you could use a `TransformedArray` from `AffineTransforms.jl`. That package will be deprecated by `CoordinateTransforms.jl`, however, but I can't remember the current status.

**Comment by tknopp** [comment]
Yes, such a syntax would be definitely a nice convenience constructor, so thumbs up.

**Comment by timholy** [comment]
Oh, and to answer an earlier question, yes, all the IO packages have been updated too if you check things out as described in https://github.com/timholy/Images.jl/issues/542#issuecomment-254059092. That gist has been updated several times, so you may want a fresh copy.

**Comment by tknopp** [comment]
Ok thanks.

Its quite interesting to see how this restructuring will change the workflow. Before, all parameters such as `pixelspacing` where just metadata and I extracted the underlaying array before giving it to a hot loop. Maybe this won't be necessary anymore if the type stability is preserved.

**Comment by timholy** [comment]
Yes, the idea is you just pass an `AbstractArray` to whatever you do. That way you never have to make choices between performance and "self-documentation." It should be just as fast as an `Array`.

Could you open an issue on ImageAxes with https://github.com/timholy/Images.jl/issues/542#issuecomment-270742133, so I don't forget?

**Comment by timholy** [comment]
Added to `AxisArrays` (need to be running master).

**Comment by timholy** [comment]
Closed by #577
